core/src/main/java/org/mule/runtime/core/internal/message/DefaultMessageBuilder.java

```diff
@@ -115,7 +115,12 @@ public DefaultMessageBuilder(org.mule.runtime.api.message.Message message) {
 
   @Override
   public InternalMessage.Builder payload(Object payload) {
-    this.payload = payload;
+    if (payload instanceof TypedValue<?>) {
+      this.payload = ((TypedValue) payload).getValue();
+      this.dataType = ((TypedValue) payload).getDataType();
```

Comment:
True, but that's beyond this PR IMO because we'd also have to avoid creating a new TypedValue  every time a new part is changed.